### ISSUE TITLE
Validation Too Strict for Custom Scalars

### ISSUE THREAD

**Issue by gaberogan** [graphql/graphiql#1705]
Custom Scalars is GraphQL's only way of providing flexible input types. As it should be, there is no validation for these in the main editor. In the variable editor however, a custom scalar is only allow to be a string, number, boolean, or null. Anything else throws an error. If I need a free form object in the variable editor, it will complain. Could we make a non-breaking change and remove validation for custom scalars here to be more consistent and flexible, allowing free form object types? (by removing `type instanceof GraphQLScalarType`)

https://github.com/graphql/graphiql/blob/fa5fe7e6f16767d74f3807ae09816733867ebf08/packages/codemirror-graphql/src/variables/lint.js#L165-L174

![Screenshot 2020-10-31 193830](https://user-images.githubusercontent.com/58148287/97789154-69efee00-1bbe-11eb-9c40-1f6472f01e75.jpg)

**Comment by acao** [comment]
hey, good question!

maybe instead it could provide a warning, such as how vscode gives a warning squiggle for `any` or implicit `any` even when `"noImplicitAny": "true"`

also of interest, there is a 2nd stage draft spec for Custom Scalar Definitions that are hosted via URIs:
https://github.com/graphql/graphql-spec/issues/635

there is this successor to the `inputUnion` spec proposal that is now RFC 1, called tagged type:
https://github.com/graphql/graphql-spec/pull/733

**Comment by gaberogan** [comment]
@acao Thanks for the input, I haven't seen those RFCs before. I think in the case of a custom scalar it would be an explicit "any" because the variable editor knows it's a custom scalar and unfortunately we can't perform any validation beyond this and have to assume it's valid.

**Comment by acao** [comment]
personally, i feel it's best to make it invalid by default until we adopt the custom scalars spec. otherwise it's a breaking change. i prefer it to be invalid by default, because `JSON`, etc custom scalars, while sometimes useful, are less than ideal (like `any` in typescript). the custom scalars spec will take care of the rest, and then GraphiQL will know how to properly validate most any custom scalar in use

**Comment by acao** [comment]
do you mind if i turn this into a feature request to support the custom scalar spec?

otherwise, it makes the most sense to mark this as a `wontfix` and close it.

**Comment by truptivishe-gep** [comment]
I too am facing this issue as one of my field is of type AnyScalarGraphType. Graphiql shows the warning :  Expected value of type "_Any". Shouldn't AnyScalarGraphType be included as a valid scalar type?

**Comment by acao** [comment]
@truptivishe-gep `Any` is not a specification scalar, and if you follow the discussion here, you'll see that there is still something in the works for providing and validating custom scalar definitions across languages

**Comment by acao** [comment]
closing this as a dupe because it's another request for validating the Any scalar

graphiql is a to-spec tool used for specification development, and closely aligned with `graphql-js`, so once the custom scalar spec lands in graphql@17 hopefully we will create a feature request to add this capability.
